src/e_ibmca.c

```diff
@@ -253,2999 +146,684 @@ static size_t size_digest_list = 0;
-		sizeof(ica_aes_vector_t) - sizeof(uint32_t),
-		EVP_CIPH_GCM_MODE | EVP_CIPH_FLAG_DEFAULT_ASN1
-		| EVP_CIPH_CUSTOM_IV | EVP_CIPH_FLAG_CUSTOM_CIPHER
-		| EVP_CIPH_ALWAYS_CALL_INIT | EVP_CIPH_CTRL_INIT
-		| EVP_CIPH_CUSTOM_COPY | EVP_CIPH_FLAG_AEAD_CIPHER,
-		sizeof(ICA_AES_GCM_CTX),
-		ibmca_aes_gcm_init_key, ibmca_aes_gcm_cipher, NULL, NULL,
-		NULL, ibmca_aes_gcm_ctrl)
-#endif
-
-DECLARE_AES_EVP(192, ecb, sizeof(ica_aes_vector_t),
-		sizeof(ica_aes_key_len_192_t), sizeof(ica_aes_vector_t),
-		EVP_CIPH_ECB_MODE, sizeof(ICA_AES_192_CTX),
-		ibmca_init_key, ibmca_aes_192_cipher, ibmca_cipher_cleanup,
-		EVP_CIPHER_set_asn1_iv, EVP_CIPHER_get_asn1_iv, NULL)
-DECLARE_AES_EVP(192, cbc, sizeof(ica_aes_vector_t),
-		sizeof(ica_aes_key_len_192_t), sizeof(ica_aes_vector_t),
-		EVP_CIPH_CBC_MODE, sizeof(ICA_AES_192_CTX),
-		ibmca_init_key, ibmca_aes_192_cipher, ibmca_cipher_cleanup,
-		EVP_CIPHER_set_asn1_iv, EVP_CIPHER_get_asn1_iv, NULL)
-DECLARE_AES_EVP(192, ofb, 1, sizeof(ica_aes_key_len_192_t),
-		sizeof(ica_aes_vector_t), EVP_CIPH_OFB_MODE,
-		sizeof(ICA_AES_192_CTX), ibmca_init_key,
-		ibmca_aes_192_cipher, ibmca_cipher_cleanup,
-		EVP_CIPHER_set_asn1_iv, EVP_CIPHER_get_asn1_iv, NULL)
-DECLARE_AES_EVP(192, cfb, 1, sizeof(ica_aes_key_len_192_t),
-		sizeof(ica_aes_vector_t), EVP_CIPH_CFB_MODE,
-		sizeof(ICA_AES_192_CTX), ibmca_init_key,
-		ibmca_aes_192_cipher, ibmca_cipher_cleanup,
-		EVP_CIPHER_set_asn1_iv, EVP_CIPHER_get_asn1_iv, NULL)
-#ifndef OPENSSL_NO_AES_GCM
-DECLARE_AES_EVP(192, gcm, 1, sizeof(ica_aes_key_len_192_t),
-		sizeof(ica_aes_vector_t) - sizeof(uint32_t),
-		EVP_CIPH_GCM_MODE | EVP_CIPH_FLAG_DEFAULT_ASN1
-		| EVP_CIPH_CUSTOM_IV | EVP_CIPH_FLAG_CUSTOM_CIPHER
-		| EVP_CIPH_ALWAYS_CALL_INIT | EVP_CIPH_CTRL_INIT
-		| EVP_CIPH_CUSTOM_COPY | EVP_CIPH_FLAG_AEAD_CIPHER,
-		sizeof(ICA_AES_GCM_CTX),
-		ibmca_aes_gcm_init_key, ibmca_aes_gcm_cipher, NULL, NULL,
-		NULL, ibmca_aes_gcm_ctrl)
-#endif
+/* Destructor (complements the "ENGINE_ibmca()" constructor) */
+static int ibmca_destroy(ENGINE *e)
+{
+    /* Unload the ibmca error strings so any error state including our
+     * functs or reasons won't lead to a segfault (they simply get displayed
+     * without corresponding string data because none will be found).
+     */
+#ifndef OLDER_OPENSSL
+    ibmca_des_ecb_destroy();
+    ibmca_des_cbc_destroy();
+    ibmca_des_ofb_destroy();
+    ibmca_des_cfb_destroy();
+    ibmca_tdes_ecb_destroy();
+    ibmca_tdes_cbc_destroy();
+    ibmca_tdes_ofb_destroy();
+    ibmca_tdes_cfb_destroy();
+
+    ibmca_aes_128_ecb_destroy();
+    ibmca_aes_128_cbc_destroy();
+    ibmca_aes_128_ofb_destroy();
+    ibmca_aes_128_cfb_destroy();
+    ibmca_aes_192_ecb_destroy();
+    ibmca_aes_192_cbc_destroy();
+    ibmca_aes_192_ofb_destroy();
+    ibmca_aes_192_cfb_destroy();
+    ibmca_aes_256_ecb_destroy();
+    ibmca_aes_256_cbc_destroy();
+    ibmca_aes_256_ofb_destroy();
+    ibmca_aes_256_cfb_destroy();
```

Comment:
implicit function declarations of all ibmca_cipher_destroy functions, must be exported via ibmca.h.